pageserver/src/tenant/timeline.rs

```diff
@@ -4235,87 +4237,147 @@ impl Timeline {
                 )))
             });
 
-            let mut wrote_keys = false;
-
-            let mut key_request_accum = KeySpaceAccum::new();
-            for range in &partition.ranges {
-                let mut key = range.start;
-                while key < range.end {
-                    // Decide whether to retain this key: usually we do, but sharded tenants may
-                    // need to drop keys that don't belong to them.  If we retain the key, add it
-                    // to `key_request_accum` for later issuing a vectored get
-                    if self.shard_identity.is_key_disposable(&key) {
-                        debug!(
-                            "Dropping key {} during compaction (it belongs on shard {:?})",
-                            key,
-                            self.shard_identity.get_shard_number(&key)
-                        );
-                    } else {
-                        key_request_accum.add_key(key);
-                    }
+            if !compact_metadata {
```

Comment:
Can you separate the metadata and non-metadata logic into different functions?

Reply:
https://github.com/neondatabase/neon/pull/7567/commits/5af19aa627fd269875e52a40ac52f0d76fb22bc8